File: src/main/java/chapter4/create/DeferExample.java
package chapter4.create;

import common.Log;
import common.Shape;
import io.reactivex.Observable;

import java.util.Arrays;
import java.util.Iterator;
import java.util.concurrent.Callable;


public class DeferExample {

    Iterator<String> colors = Arrays.asList("1", "3", "5", "6").iterator();

    public void marbleDiagram() {
        Callable<Observable<String>> supplier = () -> getObservable();
        Observable<String> source = Observable.defer(supplier);

        source.subscribe(val -> Log.it("Subscriber #1 -> " + val));
        source.subscribe(val -> Log.it("Subscriber #2 -> " + val));
    }

    public void notDeferred() {
        Observable<String> source = getObservable();
        source.subscribe(val -> Log.it("Subscriber #1 -> " + val));
        source.subscribe(val -> Log.it("Subscriber #2 -> " + val));
    }

    private Observable<String> getObservable() {
        if (colors.hasNext()) {
            String color = colors.next();
            return Observable.just(
                    Shape.getString(color, Shape.BALL),
                    Shape.getString(color, Shape.RECTANGLE),
                    Shape.getString(color, Shape.PENTAGON));
        }
        return Observable.empty();
    }

    public static void main(String[] args) {
        DeferExample demo = new DeferExample();
        demo.marbleDiagram();
        demo.notDeferred();
    }

}

File: src/main/kotlin/chapter4/DeferExample_.kt
package chapter4

import common.Log_
import common.Shape_
import io.reactivex.Observable


class DeferExample_ {

    val colors: Iterator<String> = listOf("1", "3", "5", "6").iterator()


    fun marbleDiagram() {
        val source = Observable.defer { getObservable() }

        source.subscribe { Log_.it("Subscriber #1 -> $it") }
        source.subscribe { Log_.it("Subscriber #2 -> $it") }
    }

    fun notDeferred() {
        val source = getObservable()
        source.subscribe { Log_.it("Subscriber #1 -> $it") }
        source.subscribe { Log_.it("Subscriber #2 -> $it") }
    }

    private fun getObservable(): Observable<String> =
            when (colors.hasNext()) {
                false -> Observable.empty()
                true -> {
                    val color = colors.next()
                    Observable.just(
                            Shape_.getString(color, Shape_.BALL),
                            Shape_.getString(color, Shape_.RECTANGLE),
                            Shape_.getString(color, Shape_.PENTAGON))
                }
            }

}


fun main(args: Array<String>) {
    val demo = DeferExample_()
    demo.marbleDiagram()
    demo.notDeferred()
}
